### PR TITLE
Add spi-based-sharding-spring-boot-mybatis-example for spi-based-shar…

### LINKED ISSUE #15719: [OpenForce 2022]  Add `spi-based-sharding-spring-boot-mybatis-example` for `spi-based-sharding-algorithm-example`
Hi community,

This issue is for [OpenForce 2022](https://github.com/apache/shardingsphere/discussions/12683#discussioncomment-2139441)

### Background

[spi-based-sharding-algorithm-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example) is an example of a custom algorithm that demonstrates how to integrate a user-defined sharding algorithm and configure it using  [SPI (Service Provider Interface)](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html).

### Aim

Add `spi-based-sharding-spring-boot-mybatis-example` for `spi-based-sharding-algorithm-example`

### Basic Qualifications

- Java
- Maven

### Example FYI

- [sharding-spring-boot-mybatis-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/sharding-example/sharding-spring-boot-mybatis-example)

### PR DESCRIPTION
**Pull request by susongyan**
…ding-algorithm-example #15719

Fixes #15719.

Changes proposed in this pull request:
- spi based table sharding example
- spi based database sharding example
- spi based database&table sharding example

### CODE REVIEW

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        <dependency>
+            <groupId>org.mybatis.spring.boot</groupId>
+            <artifactId>mybatis-spring-boot-starter</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (soulasuna) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml:52]: There should be a blank line at the end.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/SPIBasedShardingSpringBootMybatisExample.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.ComponentScan;
+
+import java.sql.SQLException;
+
+/**
+ * @author susongyan
+ **/
```
Comment (soulasuna) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/SPIBasedShardingSpringBootMybatisExample.java:32]: No comments needed here.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.shardingsphere.sharding.api.sharding.standard.PreciseShardingValue;
+import org.apache.shardingsphere.sharding.api.sharding.standard.RangeShardingValue;
+import org.apache.shardingsphere.sharding.api.sharding.standard.StandardShardingAlgorithm;
+
+import java.util.Collection;
+
+/**
+ * @author susongyan
+ **/
+public class SPIBasedDatasourceStandardShardingAlgorithmFixture implements StandardShardingAlgorithm<Integer> {
```
Comment (soulasuna) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java:29]: Need to add final for class.

File: pom.xml [hunk 4]
```diff
@@ -46,6 +46,9 @@
         
         <module>shardingsphere-test</module>
         <module>shardingsphere-distribution</module>
+        <module>
+            examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example
+        </module>
```
Comment (soulasuna) [inline pom.xml:51]: The `<module>` should be added in the corresponding parent pom.

**Review by soulasuna — COMMENTED** [review]
Thanks for your contributions, please fix them.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    public Collection<String> doSharding(Collection<String> availableTargetNames, RangeShardingValue<Integer> shardingValue) {
+        return availableTargetNames;
+    }
+
+    @Override
+    public void init() {
+
```
Comment (yx9o) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java:49]: Please remove unnecessary blank lines.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        return null;
+    }
+
+    private String shardingSuffix(Integer shardingValue) {
+        return "-" + (shardingValue % 2);
+    }
+
+    @Override
+    public Collection<String> doSharding(Collection<String> availableTargetNames, RangeShardingValue<Integer> shardingValue) {
```
Comment (yx9o) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/spring/boot/mybatis/fixture/SPIBasedDatasourceStandardShardingAlgorithmFixture.java:43]: Please add `final` for parameters, and check other parameters and classes.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/resources/logback.xml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<configuration>
+    <property name="log.context.name" value="sharding-spring-boot-mybatis-example" />
```
Comment (yx9o) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/src/main/resources/logback.xml:20]: Please modify `log.context.name` to the current module name.

**Comment by susongyan** [comment]
is it ready to merge now?

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            <artifactId>mybatis-spring-boot-starter</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+        </dependency>
+    </dependencies>
+
+</project>
```
Comment (RaigorJiang) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml:54]: Please keep a blank line at the end of file.

**Comment by RaigorJiang** [comment]
Hi @susongyan 
It's a good PR, there is only a small problem left to improve before merging.  

@soulasuna @yx9o  Thanks for your nice review, too.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml (hunk 8 above)
Comment (susongyan) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml]: OK，
by the way, what style check format do you use?

Comment (RaigorJiang) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml]: > OK， by the way, what style check format do you use?

Please refer to:
https://shardingsphere.apache.org/community/en/contribute/code-conduct/

Comment (susongyan) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-spring-boot-mybatis-example/pom.xml]: got it , thx

**Review by RaigorJiang — APPROVED** [review]

**Pull request merged**